hs/src/Reach/Verify/SMT.hs

```diff
@@ -1330,10 +1333,12 @@ smt_n = \case
     when um $ smtMapRefresh at
     smt_s s
   LLC_While at asn inv cond body k -> do
-    constrainedMapsRef <- asks ctxt_constrained_maps
-    liftIO $ modifyIORef' constrainedMapsRef insertConstrainedMaps
-
+    mclRef <- asks ctxt_maps_constrained_local
+    liftIO $ writeIORef mclRef S.empty
     mapM_ ctxtNewScope [before_m, loop_m, after_m]
+    mcl <- liftIO $ readIORef mclRef
+    mcaRef <- asks ctxt_maps_constrained_always
+    liftIO $ modifyIORef' mcaRef $ M.mapWithKey $ \m _ -> S.member m mcl
```

Comment:
I don't believe that this implementation works. You should make some more tests

Reply:
You're right, it doesn't. It warns about maps that cannot be constrained in a given invariant because they were defined later. I'm not sure how to glean which maps have been defined so far at a given MapReduce

Reply:
Start with more tests that demonstrate the problem. Then I think we'll need to add some sort of counter for invariants (loops) and remember that in the definition of a map, then record which whiles check each map, and then warn when one is missing a while after the one it started on